File: best_time_to_buy_and_sell_stock_ii.py
"""
122. Best Time to Buy and Sell Stock II

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        for i in range(1, len(prices)):
            if prices[i] > prices[i - 1]:
                max_profit += prices[i] - prices[i - 1]
        return max_profit
    


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    prices1 = [7, 1, 5, 3, 6, 4]
    print(f"Example 1: {solution.maxProfit(prices1)}")  # Expected: 7
    
    # Example 2
    prices2 = [1, 2, 3, 4, 5]
    print(f"Example 2: {solution.maxProfit(prices2)}")  # Expected: 4
    
    # Example 3
    prices3 = [7, 6, 4, 3, 1]
    print(f"Example 3: {solution.maxProfit(prices3)}")  # Expected: 0
    
    # Edge case: single price
    prices4 = [5]
    print(f"Example 4: {solution.maxProfit(prices4)}")  # Expected: 0
    
    # Edge case: two prices increasing
    prices5 = [1, 5]
    print(f"Example 5: {solution.maxProfit(prices5)}")  # Expected: 4
    
    # Edge case: alternating prices
    prices6 = [1, 3, 2, 4, 1, 5]
    print(f"Example 6: {solution.maxProfit(prices6)}")  # Expected: 2+2+4 = 8
    
    # All same prices
    prices7 = [5, 5, 5, 5]
    print(f"Example 7: {solution.maxProfit(prices7)}")  # Expected: 0

File: best_time_to_buy_and_sell_stock_ii.cpp
/*
122. Best Time to Buy and Sell Stock II

Time Complexity: O(n)
Space Complexity: O(1)
*/

#include <iostream>
#include <vector>
using namespace std;


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxProfit = 0;
        
        for (int i = 1; i < prices.size(); i++) {
            if (prices[i] > prices[i - 1]) {
                maxProfit += prices[i] - prices[i - 1];
            }
        }
        return maxProfit;
    }
};


// Test cases
int main() {
    Solution solution;
    
    // Example 1
    vector<int> prices1 = {7, 1, 5, 3, 6, 4};
    cout << "Example 1: " << solution.maxProfit(prices1) << endl;  // Expected: 7
    
    // Example 2
    vector<int> prices2 = {1, 2, 3, 4, 5};
    cout << "Example 2: " << solution.maxProfit(prices2) << endl;  // Expected: 4
    
    // Example 3
    vector<int> prices3 = {7, 6, 4, 3, 1};
    cout << "Example 3: " << solution.maxProfit(prices3) << endl;  // Expected: 0
    
    // Edge case: single price
    vector<int> prices4 = {5};
    cout << "Example 4: " << solution.maxProfit(prices4) << endl;  // Expected: 0
    
    // Edge case: two prices increasing
    vector<int> prices5 = {1, 5};
    cout << "Example 5: " << solution.maxProfit(prices5) << endl;  // Expected: 4
    
    // Edge case: alternating prices
    vector<int> prices6 = {1, 3, 2, 4, 1, 5};
    cout << "Example 6: " << solution.maxProfit(prices6) << endl;  // Expected: 2+2+4 = 8
    
    // All same prices
    vector<int> prices7 = {5, 5, 5, 5};
    cout << "Example 7: " << solution.maxProfit(prices7) << endl;  // Expected: 0
    
    return 0;
}
